Readme.md
# Extension ``restclient`` 

This extension basically provides a class library for sending HTTP requests and receiving HTTP responses from a RESTful API resource.
You can set many options by using Typo3-Backend, or even the HttpClient* setter.
The CRUD operations GET, POST, PUT, DELETE are supported. 
The HttpClient class is cURL library based.

## Documentation

You can find complete documentation here[1].

[1]: https://docs.typo3.org/typo3cms/extensions/restclient/

## Feedback for composer mode

I would appreciate it if you give me some feedback.
If you download and install this extension in composer mode, i please you to star my project.


ChangeLog
2020-01-03 Tommaso Sacramone  <tommasosacramone.www@gmail.com>  
  * Change vendor in namespace  
  * Release of restclient 3.0.0 in TER

2019-06-04 Tommaso Sacramone  <tommasosacramone.www@gmail.com>  
  * Fix regression for Typo3 7
  * Fix Changelog
  * Fix Settings in documentation
  * Release of restclient 2.0.6 in TER

2019-03-13 Tommaso Sacramone  <tommasosacramone.www@gmail.com>  
  * Fix regression for Typo3 7
  * Release of restclient 2.0.5 in TER

2019-02-21 Tommaso Sacramone  <tommasosacramone.www@gmail.com>  
  * Fix some deprecations for Typo3 9
  * Release of restclient 2.0.4 in TER

2017-07-28 Tommaso Sacramone  <tommasosacramone.www@gmail.com>  
  * Updated typo3 constrains to 8.7.99
  * Release of restclient 2.0.3 in TER

2017-04-24 Tommaso Sacramone  <tommasosacramone.www@gmail.com>  
  * Some small bugfix (Thanks to contributors: sebastian-seitz and to m3g4p0p)
  * Release of restclient 2.0.1 in TER

2015-02-28 Tommaso Sacramone  <tommasosacramone.www@gmail.com>  
  * Substantial changes that allow you to send data strings (eg. json)
  * Release of restclient 2.0.0 in TER

2015-12-26 Tommaso Sacramone  <tommasosacramone.www@gmail.com>  
  * Some minor improvements in HttpClient
  * Release of restclient 1.1.3 in TER

2015-11-27 Tommaso Sacramone  <tommasosacramone.www@gmail.com>  
  * Add optional feature initClient in function reset
  * Fix to integer the max_redirect field in configuration   
  * Some small updates to documentation 
  * Release of restclient 1.1.2 in TER

2015-11-23 Tommaso Sacramone  <tommasosacramone.www@gmail.com>  
  * Some small bugfix update in HttpClient Hook
  * Some small updates to documentation
  * Add composer.json, change extension icon  
  * Release of restclient 1.1.1 in TER

2015-11-16 Tommaso Sacramone  <tommasosacramone.www@gmail.com>  
  * Removed test plugin, small improvements of HttpClient code, some small bugfix update, new minor version for compatibility with TYPO3 7.6
  * Release of restclient 1.1.0 in TER

2015-11-12 Tommaso Sacramone  <tommasosacramone.www@gmail.com>  
  * Add support for User Agent, file uploads, and some helpers to HttpClientRequest
  * Release of restclient 1.0.8 in TER

2015-11-08 Tommaso Sacramone  <tommasosacramone.www@gmail.com>  
  * Add HttpClientRequest
  * Release of restclient 1.0.7 in TER

2015-11-06 Tommaso Sacramone  <tommasosacramone.www@gmail.com>  
  * Add typo3 logging, and improve documentation
  * Release of restclient 1.0.6 in TER

2015-11-03 Tommaso Sacramone  <tommasosacramone.www@gmail.com>  
  * Add max in follow redirect, header to array, and the shorthands for CRUD
  * Release of restclient 1.0.5 in TER

2015-10-30  Tommaso Sacramone  <tommasosacramone.www@gmail.com>  
  * Add hook to preprocess the response, and support cookies
  * Release of restclient 1.0.4 in TER

2015-10-28  Tommaso Sacramone  <tommasosacramone.www@gmail.com>
  * Add documentation
  * Add support https certificates, follow_redirect and check http code
  * Release of restclient 1.0.3 in TER

2015-10-26  Tommaso Sacramone  <tommasosacramone.www@gmail.com>
  * Add support authentication scheme and system http proxy 
  * Fix extension manager configuration of property 'error_throw_exception'
  * Release of restclient 1.0.2 in TER

2015-10-25  Tommaso Sacramone  <tommasosacramone.www@gmail.com>
  * Refactoring of documentation (sphinx)
  * Release of restclient 1.0.1 in TER

2015-10-23  Tommaso Sacramone  <tommasosacramone.www@gmail.com>
  * First release of restclient 1.0.0 in TER


Classes/Client/HttpClientError.php
<?php 

namespace Tommysack\Restclient\Client;

/***************************************************************
*  Copyright notice
*
*  (c) 2015 Tommaso Sacramone <tommasosacramone.www@gmail.com>
*  
*  All rights reserved
*
*  This script is part of the TYPO3 project. The TYPO3 project is
*  free software; you can redistribute it and/or modify
*  it under the terms of the GNU General Public License as published by
*  the Free Software Foundation; either version 3 of the License, or
*  (at your option) any later version.
*
*  The GNU General Public License can be found at
*  http://www.gnu.org/copyleft/gpl.html.
*
*  This script is distributed in the hope that it will be useful,
*  but WITHOUT ANY WARRANTY; without even the implied warranty of
*  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
*  GNU General Public License for more details.
*
*  This copyright notice MUST APPEAR in all copies of the script!
***************************************************************/

/**
 * HttpClient Error 
 *
 * This class is instantiated from HttpClient if an error occurs.
 */
class HttpClientError {  
  
  /**   
   * @var int The code error
   */
  protected $code;
  
  /**   
   * @var string The message error
   */
  protected $message;      
  
  /**
   * Constructor   
   */
  public function __construct($code, $message) {
    $this -> code = $code;    
    $this -> message = $message;        
  }
  
  /**
   * Get the error code
   *   
   * @return int The error code
   */
  public function getCode() {
    return $this -> code;  
  }
  
  /**
   * Get the error message
   *   
   * @return string The error message
   */
  public function getMessage() {    
    return $this -> message;  
  }  
  
}

?>


Classes/Client/HttpClientRequest.php
<?php 

namespace Tommysack\Restclient\Client;

/***************************************************************
*  Copyright notice
*
*  (c) 2015 Tommaso Sacramone <tommasosacramone.www@gmail.com>
*  
*  All rights reserved
*
*  This script is part of the TYPO3 project. The TYPO3 project is
*  free software; you can redistribute it and/or modify
*  it under the terms of the GNU General Public License as published by
*  the Free Software Foundation; either version 3 of the License, or
*  (at your option) any later version.
*
*  The GNU General Public License can be found at
*  http://www.gnu.org/copyleft/gpl.html.
*
*  This script is distributed in the hope that it will be useful,
*  but WITHOUT ANY WARRANTY; without even the implied warranty of
*  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
*  GNU General Public License for more details.
*
*  This copyright notice MUST APPEAR in all copies of the script!
***************************************************************/

/**
 * HttpClient Request 
 *
 * This class is instantiated to execute a request from HttpClient.
 */
class HttpClientRequest {
  
  const METHOD_GET = 'GET';
  const METHOD_POST = 'POST';
  const METHOD_PUT = 'PUT';
  const METHOD_DELETE = 'DELETE';
  
  /**   
   * @var string The method
   */
  protected $method;
  
  /**   
   * @var string The url
   */
  protected $url;    
  
  /**   
   * @var array The header
   */
  protected $header;    
  
  /**   
   * @var array|string The data 
   */
  protected $data;
  
  /**   
   * @var array The files
   */
  protected $files;  
  
  /**
   * Get the method
   *   
   * @return string The method
   */
  public function getMethod() {    
    return $this -> method;
  }
  
  /**
   * Set the method
   *
   * @param string The method  
   * @return HttpClientRequest The current instance
   */
  public function setMethod($method) {
    $this -> method = $method;
    return $this;  
  }
  
  /**
   * Get the url
   *   
   * @return string The url
   */
  public function getUrl() {    
    return $this -> url;
  }
  
  /**
   * Set the url
   *
   * @param string The url  
   * @return HttpClientRequest The current instance
   */
  public function setUrl($url) {
    $this -> url = $url;
    return $this;  
  }
  
  /**
   * Get the header
   *   
   * @return array The header
   */
  public function getHeader() {    
    return $this -> header;
  }
  
  /**
   * Add headers
   *
   * @param array The header  
   * @return HttpClientRequest The current instance
   */
  public function addHeaders($headers) {
    if (!isset($this -> header)) {
      $this -> header = array();
    }
    $this -> header = array_merge($this -> header, $headers);
    
    return $this;  
  }
  
  /**
   * Set the header
   *
   * @param array The header  
   * @return HttpClientRequest The current instance
   */
  public function setHeader($header) {
    $this -> header = $header;
    return $this;  
  }
  
  /**
   * Reset the header
   *   
   * @return HttpClientRequest The current instance
   */
  public function resetHeader() {
    unset($this -> header);
    return $this;  
  }

  /**
   * Get the data
   *   
   * @return array|string The data
   */
  public function getData() {    
    return $this -> data;
  }  
 
  /**
   * Reset the data
   *   
   * @return HttpClientRequest The current instance
   */
  public function resetData() {
    unset($this -> data);
    return $this;  
  }
  
  /**
   * Set the data
   *
   * @param array|string The data  
   * @return HttpClientRequest The current instance
   */
  public function setData($data) {
    $this -> data = $data;
    return $this;  
  }
  
  /**
   * Get the files
   *   
   * @return array The files
   */
  public function getFiles() {    
    return $this -> files;
  }
  
  /**
   * Add files
   *
   * @param array The files  
   * @return HttpClientRequest The current instance
   */
  public function addFiles($files) {
    if (!isset($this -> files)) {
      $this -> files = array();
    }
    $this -> files = array_merge($this -> files, $files);
    
    return $this;  
  }
  
  /**
   * Set the files
   *
   * @param array The files ex. array('file1'=> array('filename'=> 'file1.txt', 'realpath'=> '/path/to/file1', 'mimetype'=>'txt'), 'file2'=> array('filename'=> 'file2.txt', 'realpath'=> '/path/to/file2', 'mimetype'=>'txt'), ...)
   * @return HttpClientRequest The current instance
   */
  public function setFiles($files) {
    $this -> files = $files;
    return $this;  
  }
  
  /**
   * Reset the files
   *   
   * @return HttpClientRequest The current instance
   */
  public function resetFiles() {
    unset($this -> files);
    return $this;  
  }  
   
}

?>


Classes/Client/HttpClientResponse.php
<?php 

namespace Tommysack\Restclient\Client;

/***************************************************************
*  Copyright notice
*
*  (c) 2015 Tommaso Sacramone <tommasosacramone.www@gmail.com>
*  
*  All rights reserved
*
*  This script is part of the TYPO3 project. The TYPO3 project is
*  free software; you can redistribute it and/or modify
*  it under the terms of the GNU General Public License as published by
*  the Free Software Foundation; either version 3 of the License, or
*  (at your option) any later version.
*
*  The GNU General Public License can be found at
*  http://www.gnu.org/copyleft/gpl.html.
*
*  This script is distributed in the hope that it will be useful,
*  but WITHOUT ANY WARRANTY; without even the implied warranty of
*  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
*  GNU General Public License for more details.
*
*  This copyright notice MUST APPEAR in all copies of the script!
***************************************************************/

/**
 * HttpClient Response 
 *
 * This class is instantiated and setting from HttpClient as response of a request.
 */
class HttpClientResponse {  
  
  /**   
   * @var string The header
   */
  protected $header;
  
  /**   
   * @var string The body
   */
  protected $body;    
  
  /**   
   * @var int The request size
   */
  protected $requestSize;    
  
  /**   
   * @var string The content type
   */
  protected $contentType;
  
  /**   
   * @var int The http code
   */
  protected $httpCode;
  
  /**   
   * @var float The total time
   */
  protected $totalTime;
  
  /**   
   * @var float The connection time
   */
  protected $connectTime;
  
  /**   
   * @var string The primary ip
   */
  protected $primaryIp;
  
  /**   
   * @var string The primary port
   */
  protected $primaryPort;
  
  /**
   * Constructor   
   */
  public function __construct($header, $body, $requestSize, $contentType, $httpCode, $totalTime, $connectTime, $primaryIp, $primaryPort) {
    $this -> header = $header;    
    $this -> body = $body;
    $this -> requestSize = $requestSize;
    $this -> contentType = $contentType;
    $this -> httpCode = $httpCode;
    $this -> totalTime = $totalTime;
    $this -> connectTime = $connectTime;
    $this -> primaryIp = $primaryIp;
    $this -> primaryPort = $primaryPort;
  }
  
  /**
   * Get the header
   *   
   * @return mixed The header
   */
  public function getHeader($toArray = false) {
    $header = $this -> header;    
    if (isset($toArray) && $toArray === true) {
      $headerToarray = array();
      foreach (explode("\r\n", $header) as $index => $row) {  
        if ($index === 0) {
          $headerToArray['Status-Code'] = $row;
        }
        else {          
          list ($k, $v) = explode(': ', $row);
          $headerToArray[$k] = $v;          
        }        
      }
      
      $header = $headerToArray;
    }
    return $header;  
  }
  
  /**
   * Get the body
   *
   * @return mixed The body
   */
  public function getBody() {
    $body = $this -> body;    
    return $body;  
  }
  
  /**
   * Get the request size
   *   
   * @return int The request size
   */
  public function getRequestSize() {
    return $this -> requestSize;  
  }
  
  /**
   * Get the content type
   *   
   * @return string The Content Type
   */
  public function getContentType() {
    return $this -> contentType;  
  }
  
  /**
   * Get the http code
   *   
   * @return int The http code
   */
  public function getHttpCode() {
    return $this -> httpCode;  
  }
  
  /**
   * Get the total time
   *   
   * @return float The total time
   */
  public function getTotalTime() {
    return $this -> totalTime;  
  }
  
  /**
   * Get the connect time
   *   
   * @return float The connect time
   */
  public function getConnectTime() {
    return $this -> connectTime;  
  }
  
  /**
   * Get the primary ip
   *   
   * @return string The primary ip
   */
  public function getPrimaryIp() {
    return $this -> primaryIp;  
  }
  
  /**
   * Get the primary port
   *   
   * @return string The primary port
   */
  public function getPrimaryPort() {
    return $this -> primaryPort;  
  }
  
  
   
}

?>


Classes/Client/HttpClientException.php
<?php 

namespace Tommysack\Restclient\Client;

/***************************************************************
*  Copyright notice
*
*  (c) 2015 Tommaso Sacramone <tommasosacramone.www@gmail.com>
*  
*  All rights reserved
*
*  This script is part of the TYPO3 project. The TYPO3 project is
*  free software; you can redistribute it and/or modify
*  it under the terms of the GNU General Public License as published by
*  the Free Software Foundation; either version 3 of the License, or
*  (at your option) any later version.
*
*  The GNU General Public License can be found at
*  http://www.gnu.org/copyleft/gpl.html.
*
*  This script is distributed in the hope that it will be useful,
*  but WITHOUT ANY WARRANTY; without even the implied warranty of
*  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
*  GNU General Public License for more details.
*
*  This copyright notice MUST APPEAR in all copies of the script!
***************************************************************/

/**
 * HttpClient Exception 
 *
 * This class is exception thrown from HttpClient if an error occurs.
 */
class HttpClientException extends \Exception {  

  public function __construct($message, $code = 0, Exception $previous = null) {
    parent::__construct($message, $code, $previous);
  }  
}

?>


ext_emconf.php
<?php

########################################################################
# Extension Manager/Repository config file for ext "restclient".
#
# Auto generated 13-10-2015 19:25
#
# Manual updates:
# Only the data in the array - everything else is removed by next
# writing. "version" and "dependencies" must not be touched!
########################################################################

$EM_CONF[$_EXTKEY] = array(
  'title' => 'REST client',
  'description' => 'A client to consume RESTful Web Services',
  'category' => 'misc',
  'author' => 'Tommaso Sacramone',
  'author_email' => 'tommasosacramone.www@gmail.com',
  'shy' => '',
  'dependencies' => 'php,typo3',
  'conflicts' => '',
  'priority' => '',
  'module' => '',
  'state' => 'beta',
  'internal' => '',
  'uploadfolder' => 0,
  'createDirs' => '',
  'modify_tables' => '',
  'clearCacheOnLoad' => 0,
  'lockType' => '',
  'author_company' => '',
  'version' => '2.0.6',
  'constraints' => array(
    'depends' => array(      
      'typo3' => '6.2.0-9.9.99',
    ),
    'conflicts' => array(
    ),
    'suggests' => array(
    ),
  ),
  '_md5_values_when_last_written' => 'a:5:{s:9:"ChangeLog";s:4:"d096";s:12:"ext_icon.gif";s:4:"1bdc";s:10:"README.txt";s:4:"ee2d";s:19:"doc/wizard_form.dat";s:4:"afdf";s:20:"doc/wizard_form.html";s:4:"7145";}',
);

?>
